File: study/examples/Algorithm/c/21_number_theory.c
/*
 * 정수론 (Number Theory)
 * GCD/LCM, 소수, 모듈러 연산, 조합론
 *
 * 수학적 알고리즘의 기초입니다.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>

#define MOD 1000000007
#define MAX_N 100001

/* =============================================================================
 * 1. GCD / LCM
 * ============================================================================= */

long long gcd(long long a, long long b) {
    while (b) {
        long long temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

long long gcd_recursive(long long a, long long b) {
    if (b == 0) return a;
    return gcd_recursive(b, a % b);
}

long long lcm(long long a, long long b) {
    return a / gcd(a, b) * b;
}

/* 확장 유클리드 알고리즘: ax + by = gcd(a, b) */
long long extended_gcd(long long a, long long b, long long* x, long long* y) {
    if (b == 0) {
        *x = 1;
        *y = 0;
        return a;
    }
    long long x1, y1;
    long long g = extended_gcd(b, a % b, &x1, &y1);
    *x = y1;
    *y = x1 - (a / b) * y1;
    return g;
}

/* =============================================================================
 * 2. 소수 판별 및 체
 * ============================================================================= */

bool is_prime(long long n) {
    if (n < 2) return false;
    if (n == 2) return true;
    if (n % 2 == 0) return false;
    for (long long i = 3; i * i <= n; i += 2) {
        if (n % i == 0) return false;
    }
    return true;
}

/* 에라토스테네스의 체 */
bool* sieve_of_eratosthenes(int n) {
    bool* is_prime_arr = malloc((n + 1) * sizeof(bool));
    memset(is_prime_arr, true, n + 1);
    is_prime_arr[0] = is_prime_arr[1] = false;

    for (int i = 2; i * i <= n; i++) {
        if (is_prime_arr[i]) {
            for (int j = i * i; j <= n; j += i) {
                is_prime_arr[j] = false;
            }
        }
    }
    return is_prime_arr;
}

/* 선형 체 (소수 목록 생성) */
int* linear_sieve(int n, int* prime_count) {
    int* spf = calloc(n + 1, sizeof(int));  /* smallest prime factor */
    int* primes = malloc((n + 1) * sizeof(int));
    *prime_count = 0;

    for (int i = 2; i <= n; i++) {
        if (spf[i] == 0) {
            spf[i] = i;
            primes[(*prime_count)++] = i;
        }
        for (int j = 0; j < *prime_count && primes[j] <= spf[i] && i * primes[j] <= n; j++) {
            spf[i * primes[j]] = primes[j];
        }
    }

    free(spf);
    return primes;
}

/* 소인수분해 */
typedef struct {
    int prime;
    int count;
} PrimeFactor;

int factorize(long long n, PrimeFactor factors[]) {
    int count = 0;
    for (long long d = 2; d * d <= n; d++) {
        if (n % d == 0) {
            factors[count].prime = d;
            factors[count].count = 0;
            while (n % d == 0) {
                factors[count].count++;
                n /= d;
            }
            count++;
        }
    }
    if (n > 1) {
        factors[count].prime = n;
        factors[count].count = 1;
        count++;
    }
    return count;
}

/* =============================================================================
 * 3. 모듈러 연산
 * ============================================================================= */

long long mod_add(long long a, long long b, long long mod) {
    return ((a % mod) + (b % mod)) % mod;
}

long long mod_sub(long long a, long long b, long long mod) {
    return ((a % mod) - (b % mod) + mod) % mod;
}

long long mod_mul(long long a, long long b, long long mod) {
    return ((a % mod) * (b % mod)) % mod;
}

/* 빠른 거듭제곱 */
long long mod_pow(long long base, long long exp, long long mod) {
    long long result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp & 1) {
            result = mod_mul(result, base, mod);
        }
        base = mod_mul(base, base, mod);
        exp >>= 1;
    }
    return result;
}

/* 모듈러 역원 (페르마의 소정리, mod가 소수일 때) */
long long mod_inverse(long long a, long long mod) {
    return mod_pow(a, mod - 2, mod);
}

/* 모듈러 역원 (확장 유클리드) */
long long mod_inverse_ext(long long a, long long mod) {
    long long x, y;
    long long g = extended_gcd(a, mod, &x, &y);
    if (g != 1) return -1;  /* 역원 없음 */
    return (x % mod + mod) % mod;
}

/* 모듈러 나눗셈 */
long long mod_div(long long a, long long b, long long mod) {
    return mod_mul(a, mod_inverse(b, mod), mod);
}

/* =============================================================================
 * 4. 조합론 (Combinatorics)
 * ============================================================================= */

long long factorial[MAX_N];
long long inv_factorial[MAX_N];

void precompute_factorials(int n, long long mod) {
    factorial[0] = 1;
    for (int i = 1; i <= n; i++) {
        factorial[i] = mod_mul(factorial[i - 1], i, mod);
    }
    inv_factorial[n] = mod_inverse(factorial[n], mod);
    for (int i = n - 1; i >= 0; i--) {
        inv_factorial[i] = mod_mul(inv_factorial[i + 1], i + 1, mod);
    }
}

/* nCr (mod p) */
long long nCr(int n, int r, long long mod) {
    if (r < 0 || r > n) return 0;
    return mod_mul(factorial[n], mod_mul(inv_factorial[r], inv_factorial[n - r], mod), mod);
}

/* nPr (mod p) */
long long nPr(int n, int r, long long mod) {
    if (r < 0 || r > n) return 0;
    return mod_mul(factorial[n], inv_factorial[n - r], mod);
}

/* 파스칼의 삼각형 (작은 n용) */
long long** pascal_triangle(int n) {
    long long** C = malloc((n + 1) * sizeof(long long*));
    for (int i = 0; i <= n; i++) {
        C[i] = calloc(n + 1, sizeof(long long));
        C[i][0] = 1;
        for (int j = 1; j <= i; j++) {
            C[i][j] = (C[i - 1][j - 1] + C[i - 1][j]) % MOD;
        }
    }
    return C;
}

/* =============================================================================
 * 5. 오일러 피 함수
 * ============================================================================= */

long long euler_phi(long long n) {
    long long result = n;
    for (long long p = 2; p * p <= n; p++) {
        if (n % p == 0) {
            while (n % p == 0) n /= p;
            result -= result / p;
        }
    }
    if (n > 1) result -= result / n;
    return result;
}

/* 오일러 피 체 */
int* euler_phi_sieve(int n) {
    int* phi = malloc((n + 1) * sizeof(int));
    for (int i = 0; i <= n; i++) phi[i] = i;

    for (int i = 2; i <= n; i++) {
        if (phi[i] == i) {  /* i is prime */
            for (int j = i; j <= n; j += i) {
                phi[j] -= phi[j] / i;
            }
        }
    }
    return phi;
}

/* =============================================================================
 * 6. 중국인의 나머지 정리 (CRT)
 * ============================================================================= */

/* x ≡ a1 (mod m1), x ≡ a2 (mod m2) */
long long crt_two(long long a1, long long m1, long long a2, long long m2) {
    long long x, y;
    long long g = extended_gcd(m1, m2, &x, &y);

    if ((a2 - a1) % g != 0) return -1;  /* 해 없음 */

    long long lcm_val = m1 / g * m2;
    long long result = a1 + m1 * ((a2 - a1) / g % (m2 / g) * x % (m2 / g));
    return ((result % lcm_val) + lcm_val) % lcm_val;
}

/* =============================================================================
 * 7. 약수 관련
 * ============================================================================= */

int count_divisors(long long n) {
    int count = 0;
    for (long long i = 1; i * i <= n; i++) {
        if (n % i == 0) {
            count++;
            if (i != n / i) count++;
        }
    }
    return count;
}

long long sum_divisors(long long n) {
    long long sum = 0;
    for (long long i = 1; i * i <= n; i++) {
        if (n % i == 0) {
            sum += i;
            if (i != n / i) sum += n / i;
        }
    }
    return sum;
}

int* get_divisors(long long n, int* count) {
    int* divisors = malloc(10000 * sizeof(int));
    *count = 0;
    for (long long i = 1; i * i <= n; i++) {
        if (n % i == 0) {
            divisors[(*count)++] = i;
            if (i != n / i) divisors[(*count)++] = n / i;
        }
    }
    return divisors;
}

/* =============================================================================
 * 테스트
 * ============================================================================= */

int compare_int(const void* a, const void* b) {
    return *(int*)a - *(int*)b;
}

int main(void) {
    printf("============================================================\n");
    printf("정수론 (Number Theory) 예제\n");
    printf("============================================================\n");

    /* 1. GCD / LCM */
    printf("\n[1] GCD / LCM\n");
    printf("    gcd(48, 18) = %lld\n", gcd(48, 18));
    printf("    lcm(12, 18) = %lld\n", lcm(12, 18));

    long long x, y;
    long long g = extended_gcd(35, 15, &x, &y);
    printf("    35*%lld + 15*%lld = %lld (gcd)\n", x, y, g);

    /* 2. 소수 */
    printf("\n[2] 소수 판별 및 체\n");
    printf("    is_prime(17) = %s\n", is_prime(17) ? "true" : "false");
    printf("    is_prime(18) = %s\n", is_prime(18) ? "true" : "false");

    int prime_count;
    int* primes = linear_sieve(50, &prime_count);
    printf("    50 이하 소수 (%d개): ", prime_count);
    for (int i = 0; i < prime_count; i++) printf("%d ", primes[i]);
    printf("\n");
    free(primes);

    /* 3. 소인수분해 */
    printf("\n[3] 소인수분해\n");
    PrimeFactor factors[20];
    int factor_count = factorize(360, factors);
    printf("    360 = ");
    for (int i = 0; i < factor_count; i++) {
        printf("%d^%d", factors[i].prime, factors[i].count);
        if (i < factor_count - 1) printf(" × ");
    }
    printf("\n");

    /* 4. 모듈러 연산 */
    printf("\n[4] 모듈러 연산\n");
    printf("    2^10 mod 1000 = %lld\n", mod_pow(2, 10, 1000));
    printf("    3^(-1) mod 7 = %lld\n", mod_inverse(3, 7));
    printf("    검증: 3 × %lld mod 7 = %lld\n", mod_inverse(3, 7),
           mod_mul(3, mod_inverse(3, 7), 7));

    /* 5. 조합론 */
    printf("\n[5] 조합론\n");
    precompute_factorials(1000, MOD);
    printf("    10! = %lld\n", factorial[10]);
    printf("    C(10, 3) = %lld\n", nCr(10, 3, MOD));
    printf("    P(10, 3) = %lld\n", nPr(10, 3, MOD));

    /* 6. 오일러 피 */
    printf("\n[6] 오일러 피 함수\n");
    printf("    φ(12) = %lld\n", euler_phi(12));
    printf("    φ(13) = %lld (소수)\n", euler_phi(13));

    /* 7. 약수 */
    printf("\n[7] 약수\n");
    printf("    28의 약수 개수: %d\n", count_divisors(28));
    printf("    28의 약수 합: %lld\n", sum_divisors(28));
    int div_count;
    int* divisors = get_divisors(28, &div_count);
    qsort(divisors, div_count, sizeof(int), compare_int);
    printf("    28의 약수: ");
    for (int i = 0; i < div_count; i++) printf("%d ", divisors[i]);
    printf("\n");
    free(divisors);

    /* 8. 복잡도 */
    printf("\n[8] 복잡도\n");
    printf("    | 알고리즘        | 시간복잡도    |\n");
    printf("    |-----------------|---------------|\n");
    printf("    | GCD (유클리드)  | O(log min)    |\n");
    printf("    | 소수 판별       | O(√n)         |\n");
    printf("    | 에라토스테네스  | O(n log log n)|\n");
    printf("    | 소인수분해      | O(√n)         |\n");
    printf("    | 거듭제곱        | O(log exp)    |\n");

    printf("\n============================================================\n");

    return 0;
}

File: study/examples/Algorithm/python/21_number_theory.py
"""
수학과 정수론 (Number Theory)
Number Theory Algorithms

정수론의 기본 알고리즘을 구현합니다.
"""

from typing import List, Tuple
from functools import reduce
import math


# =============================================================================
# 1. GCD / LCM
# =============================================================================

def gcd(a: int, b: int) -> int:
    """
    최대공약수 (유클리드 호제법)
    시간복잡도: O(log(min(a, b)))
    """
    while b:
        a, b = b, a % b
    return a


def lcm(a: int, b: int) -> int:
    """최소공배수"""
    return a * b // gcd(a, b)


def gcd_multiple(numbers: List[int]) -> int:
    """여러 수의 GCD"""
    return reduce(gcd, numbers)


def lcm_multiple(numbers: List[int]) -> int:
    """여러 수의 LCM"""
    return reduce(lcm, numbers)


def extended_gcd(a: int, b: int) -> Tuple[int, int, int]:
    """
    확장 유클리드 알고리즘
    gcd(a, b) = a*x + b*y를 만족하는 (gcd, x, y) 반환
    """
    if b == 0:
        return a, 1, 0

    g, x1, y1 = extended_gcd(b, a % b)
    x = y1
    y = x1 - (a // b) * y1

    return g, x, y


# =============================================================================
# 2. 소수 (Prime Numbers)
# =============================================================================

def is_prime(n: int) -> bool:
    """
    소수 판별
    시간복잡도: O(√n)
    """
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False

    for i in range(3, int(n ** 0.5) + 1, 2):
        if n % i == 0:
            return False
    return True


def sieve_of_eratosthenes(n: int) -> List[int]:
    """
    에라토스테네스의 체
    n 이하의 모든 소수 반환
    시간복잡도: O(n log log n)
    """
    if n < 2:
        return []

    is_prime_arr = [True] * (n + 1)
    is_prime_arr[0] = is_prime_arr[1] = False

    for i in range(2, int(n ** 0.5) + 1):
        if is_prime_arr[i]:
            for j in range(i * i, n + 1, i):
                is_prime_arr[j] = False

    return [i for i in range(n + 1) if is_prime_arr[i]]


def prime_factorization(n: int) -> List[Tuple[int, int]]:
    """
    소인수분해
    반환: [(소수, 지수), ...]
    시간복잡도: O(√n)
    """
    factors = []
    d = 2

    while d * d <= n:
        if n % d == 0:
            count = 0
            while n % d == 0:
                n //= d
                count += 1
            factors.append((d, count))
        d += 1

    if n > 1:
        factors.append((n, 1))

    return factors


# =============================================================================
# 3. 모듈러 연산 (Modular Arithmetic)
# =============================================================================

def mod_pow(base: int, exp: int, mod: int) -> int:
    """
    빠른 거듭제곱 (모듈러)
    시간복잡도: O(log exp)
    """
    result = 1
    base %= mod

    while exp > 0:
        if exp % 2 == 1:
            result = (result * base) % mod
        exp //= 2
        base = (base * base) % mod

    return result


def mod_inverse(a: int, mod: int) -> int:
    """
    모듈러 역원 (mod가 소수일 때)
    a^(-1) mod p = a^(p-2) mod p (페르마 소정리)
    """
    return mod_pow(a, mod - 2, mod)


def mod_inverse_extended(a: int, mod: int) -> int:
    """
    모듈러 역원 (확장 유클리드)
    gcd(a, mod) = 1일 때만 존재
    """
    g, x, _ = extended_gcd(a, mod)
    if g != 1:
        return -1  # 역원 없음
    return x % mod


# =============================================================================
# 4. 조합론 (Combinatorics)
# =============================================================================

def factorial_mod(n: int, mod: int) -> int:
    """n! mod p"""
    result = 1
    for i in range(2, n + 1):
        result = (result * i) % mod
    return result


class Combination:
    """
    조합 계산 (모듈러)
    전처리: O(n)
    쿼리: O(1)
    """

    def __init__(self, n: int, mod: int):
        self.mod = mod
        self.fact = [1] * (n + 1)
        self.inv_fact = [1] * (n + 1)

        # 팩토리얼 계산
        for i in range(1, n + 1):
            self.fact[i] = self.fact[i - 1] * i % mod

        # 역 팩토리얼 계산
        self.inv_fact[n] = mod_pow(self.fact[n], mod - 2, mod)
        for i in range(n - 1, -1, -1):
            self.inv_fact[i] = self.inv_fact[i + 1] * (i + 1) % mod

    def nCr(self, n: int, r: int) -> int:
        """nCr mod p"""
        if r < 0 or r > n:
            return 0
        return self.fact[n] * self.inv_fact[r] % self.mod * self.inv_fact[n - r] % self.mod

    def nPr(self, n: int, r: int) -> int:
        """nPr mod p"""
        if r < 0 or r > n:
            return 0
        return self.fact[n] * self.inv_fact[n - r] % self.mod


# =============================================================================
# 5. 오일러 피 함수 (Euler's Totient)
# =============================================================================

def euler_phi(n: int) -> int:
    """
    φ(n) = n보다 작거나 같고 n과 서로소인 양의 정수의 개수
    시간복잡도: O(√n)
    """
    result = n

    d = 2
    while d * d <= n:
        if n % d == 0:
            while n % d == 0:
                n //= d
            result -= result // d
        d += 1

    if n > 1:
        result -= result // n

    return result


def euler_phi_sieve(n: int) -> List[int]:
    """1~n까지의 오일러 피 함수 값 (체)"""
    phi = list(range(n + 1))

    for i in range(2, n + 1):
        if phi[i] == i:  # i가 소수
            for j in range(i, n + 1, i):
                phi[j] -= phi[j] // i

    return phi


# =============================================================================
# 6. 중국인의 나머지 정리 (CRT)
# =============================================================================

def chinese_remainder_theorem(remainders: List[int], moduli: List[int]) -> int:
    """
    x ≡ r_i (mod m_i)를 만족하는 최소 양의 정수 x
    모든 m_i는 서로소여야 함
    """
    M = 1
    for m in moduli:
        M *= m

    result = 0
    for r, m in zip(remainders, moduli):
        Mi = M // m
        yi = mod_inverse_extended(Mi, m)
        result += r * Mi * yi

    return result % M


# =============================================================================
# 7. 이항 계수 (Lucas Theorem)
# =============================================================================

def lucas(n: int, r: int, p: int) -> int:
    """
    루카스 정리로 nCr mod p 계산
    p가 소수일 때, n, r이 매우 클 때 유용
    """
    if r == 0:
        return 1
    return lucas(n // p, r // p, p) * nCr_small(n % p, r % p, p) % p


def nCr_small(n: int, r: int, p: int) -> int:
    """작은 nCr mod p"""
    if r > n:
        return 0
    if r == 0 or r == n:
        return 1

    num = den = 1
    for i in range(r):
        num = num * (n - i) % p
        den = den * (i + 1) % p

    return num * mod_pow(den, p - 2, p) % p


# =============================================================================
# 8. 약수 (Divisors)
# =============================================================================

def divisors(n: int) -> List[int]:
    """
    n의 모든 약수
    시간복잡도: O(√n)
    """
    result = []

    for i in range(1, int(n ** 0.5) + 1):
        if n % i == 0:
            result.append(i)
            if i != n // i:
                result.append(n // i)

    return sorted(result)


def divisor_count(n: int) -> int:
    """약수의 개수"""
    factors = prime_factorization(n)
    count = 1
    for _, exp in factors:
        count *= (exp + 1)
    return count


def divisor_sum(n: int) -> int:
    """약수의 합"""
    factors = prime_factorization(n)
    total = 1
    for p, e in factors:
        total *= (pow(p, e + 1) - 1) // (p - 1)
    return total


# =============================================================================
# 9. 선형 디오판토스 방정식
# =============================================================================

def solve_linear_diophantine(a: int, b: int, c: int) -> Tuple[bool, int, int]:
    """
    ax + by = c의 정수해 (x, y) 찾기
    반환: (해 존재 여부, x, y)
    """
    g, x0, y0 = extended_gcd(abs(a), abs(b))

    if c % g != 0:
        return False, 0, 0

    x0 *= c // g
    y0 *= c // g

    if a < 0:
        x0 = -x0
    if b < 0:
        y0 = -y0

    return True, x0, y0


# =============================================================================
# 테스트
# =============================================================================

def main():
    print("=" * 60)
    print("수학과 정수론 (Number Theory) 예제")
    print("=" * 60)

    # 1. GCD / LCM
    print("\n[1] GCD / LCM")
    a, b = 48, 18
    print(f"    gcd({a}, {b}) = {gcd(a, b)}")
    print(f"    lcm({a}, {b}) = {lcm(a, b)}")
    g, x, y = extended_gcd(a, b)
    print(f"    확장 유클리드: {a}*{x} + {b}*{y} = {g}")

    # 2. 소수
    print("\n[2] 소수")
    print(f"    17 is prime: {is_prime(17)}")
    print(f"    18 is prime: {is_prime(18)}")
    primes = sieve_of_eratosthenes(50)
    print(f"    50 이하 소수: {primes}")
    print(f"    60의 소인수분해: {prime_factorization(60)}")

    # 3. 모듈러 연산
    print("\n[3] 모듈러 연산")
    mod = 1000000007
    print(f"    2^10 mod {mod} = {mod_pow(2, 10, mod)}")
    print(f"    3의 역원 mod 7 = {mod_inverse(3, 7)}")
    print(f"    검증: 3 * {mod_inverse(3, 7)} mod 7 = {3 * mod_inverse(3, 7) % 7}")

    # 4. 조합
    print("\n[4] 조합론")
    comb = Combination(1000, mod)
    print(f"    10C3 = {comb.nCr(10, 3)}")
    print(f"    10P3 = {comb.nPr(10, 3)}")
    print(f"    100C50 mod {mod} = {comb.nCr(100, 50)}")

    # 5. 오일러 피 함수
    print("\n[5] 오일러 피 함수")
    for n in [10, 12, 7, 1]:
        print(f"    φ({n}) = {euler_phi(n)}")

    # 6. 중국인의 나머지 정리
    print("\n[6] 중국인의 나머지 정리")
    remainders = [2, 3, 2]
    moduli = [3, 5, 7]
    x = chinese_remainder_theorem(remainders, moduli)
    print(f"    x ≡ 2 (mod 3), x ≡ 3 (mod 5), x ≡ 2 (mod 7)")
    print(f"    x = {x}")
    print(f"    검증: {x % 3}, {x % 5}, {x % 7}")

    # 7. 루카스 정리
    print("\n[7] 루카스 정리")
    n, r, p = 1000, 500, 13
    print(f"    C({n}, {r}) mod {p} = {lucas(n, r, p)}")

    # 8. 약수
    print("\n[8] 약수")
    n = 36
    print(f"    {n}의 약수: {divisors(n)}")
    print(f"    약수 개수: {divisor_count(n)}")
    print(f"    약수 합: {divisor_sum(n)}")

    # 9. 디오판토스 방정식
    print("\n[9] 선형 디오판토스 방정식")
    a, b, c = 3, 5, 7
    exists, x, y = solve_linear_diophantine(a, b, c)
    print(f"    {a}x + {b}y = {c}")
    print(f"    해 존재: {exists}, x={x}, y={y}")
    print(f"    검증: {a}*{x} + {b}*{y} = {a * x + b * y}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()
